### ISSUE TITLE
Consider accepting non-inexact primals in jax.vjp (like jax.grad)

### ISSUE THREAD

**Issue by NeilGirdhar** [google/jax#4308]
Edit: See most recent comments for actual issue description 

----

Is it possible to define a custom VJP for a function with respect to some outputs, but not others?  In the following synthetic example, I want to ignore the cotangents with respect to the RNG.  JAX rightly tells me I won't get any cotangents for it ("Primal inputs to reverse-mode differentiation must be of float or complex type, got type uint32")  Yeah, fair enough, but I want to tell JAX that I don't need them.
```python
from functools import partial
from jax.random import normal, split, PRNGKey
from jax import custom_vjp, vjp


@partial(custom_vjp, nondiff_argnums=(1,))
def f(x, rng):
    rng, new_rng = split(rng)
    y = normal(rng)
    return x + y, new_rng


def f_fwd(x, rng):
    return f(x, rng), None


def f_bwd(residuals, cotangents):
    x_bar, rng_bar = cotangents     # I just want x_bar here!
    return x_bar,


f.defvjp(f_fwd, f_bwd)


rng = PRNGKey(0)
z, f_vjp = vjp(f, 5.0, rng)
f_vjp(1.0)  # I just want to pass in x_bar here!
```
Maybe if `custom_vjp` had an additional parameter?  Either:
- `has_aux` to indicate that the function returns a pair, the first element being differentiable, the second being auxilliary,
- `nondiff_outputs` to indicate that the function outputs a tuple, and this parameter's value indicating which elements of the tuple are nondifferentiable, or
- `nondiff_output_f`, a function that selects outputs from `f` and returns a pytree that is differentiable.  JAX can call this function on the cotangent structure that matches the output of f, and produce something for the backward pass.

The above example would be resolved with `has_aux=True` or `nondiff_outputs=0` or `nondiff_output_f=lambda z: z[0]`.

In all three cases, JAX could just do its type check on the extracted cotangents (that are fed to the backward pass) rather than the output of the forward pass.  As long as those are real or complex, then I think JAX should allow it.

**Comment by shoyer** [comment]
I think this just works for custom VJP, as long as you don't try to differentiate with respect to any of the non-differentiable arguments?

Extending your previous example:
```python
from functools import partial
from jax.random import normal, split, PRNGKey
from jax import custom_vjp, grad

@partial(custom_vjp, nondiff_argnums=(1,))
def f(x, rng):
    rng, new_rng = split(rng)
    y = normal(rng)
    return x + y, new_rng

def f_fwd(x, rng):
    return f(x, rng), None

def f_bwd(aux, residuals, cotangents):
    x_bar, rng_bar = cotangents 
    return x_bar,

f.defvjp(f_fwd, f_bwd)

def objective(x, rng):
  return f(x, rng)[0]

rng = PRNGKey(0)
grad(objective)(5.0, rng)  # by default, differentiates with respect to the argument 0
```

You'd need to do something a _little_ different for `jax.vjp`, but hopefully that shows the general idea...

**Comment by NeilGirdhar** [comment]
Ohhh!  I assumed that when I got that error that JAX wouldn't let me ever pass have an integer cotangent, but I guess it's just at the point at which I request the gradient that matters.  Thanks for taking the time to set me straight!

**Comment by NeilGirdhar** [comment]
@shoyer Is there a way to make the last line work using `vjp`?  I thought that `grad(objective)(*primals)` was just `vjp(*primals)[1](identity)` where identity is the appropriate pytree?

**Comment by shoyer** [comment]
> @shoyer Is there a way to make the last line work using `vjp`? I thought that `grad(objective)(*primals)` was just `vjp(objective, *primals)[1](1.0)`?

You can do this with a `partial` (or closure) to avoid passing the integer argument in `*primals`, e.g.,
```
>>> vjp(partial(objective, rng=rng), 5.0)[1](1.0)
(1.0,)
```

**Comment by NeilGirdhar** [comment]
Interesting, thank you!  Would you be able to provide some insight into why the primals of VJP have be inexact types, but those of grad don't?

**Comment by shoyer** [comment]
> Interesting, thank you! Would you be able to provide some insight into why the primals of VJP have be inexact types, but those of grad don't?

`grad` is basically a convenience wrapper around the lower-level `vjp`. Every argument to a `grad` decorated function that isn't listed in `argnums` is partially applied before calling `vjp`. If you look at the implementation of `value_and_grad` in `api.py`, it basically does exactly this with `argnums_partial`.

There's no reason why `vjp` _couldn't_ have the `argnums` argument, but I guess it seemed unnecessarily complex for a lower level function.

**Comment by NeilGirdhar** [comment]
Okay, thanks.  I need to pass an aggregate structure into VJP in my fixed point library, so unfortunate I can't use your closure idea.  I'm going to try to build a shim around `vjp` so that it just works in this case.

It seems like both `jax.vjp` and `jax.grad` now use `_vjp` as their low level function, so I guess theoretically, `vjp` could support non-inexact primals.  On the bright side, at least `jax.argnums_partial` is exposed  (though undocumented).

**Comment by NeilGirdhar** [comment]
Oh, never mind, I misunderstood.  `argnums` doesn't help me; I would need a generalized version of that.  

I have a method `extract_comparand` that maps a `State` to a `Comparand`, which is a subset of elements.  I want some kind of generalized argnums that can look at these two structures and figure out which parts of the state are not in comparand and mark those as nondifferentiable.  Does that exist deep in JAX somewhere?

The easy solution if not is for the user to provide another function that essentially undoes `extract_comparand`, so that I can take the `Comparand`-typed cotangents and implant those into a `State` to form `State`-typed cotangents that I need.

**Comment by shoyer** [comment]
> I'm also going to try to build a shim around `vjp` so that it just works in this case. On the bright side, at least `jax.argnums_partial` is exposed (though undocumented).

Note that `argnums_partial` is not a public JAX API. If you use undocumented functions like this, you are likely to be broken eventually :)

I suspect writing your a shim around `vjp` would be appropriate for your use-case. I don't know of any existing machinery for doing the sort of comparison you describe. The general pattern in JAX is keep non-differentiable arguments entirely separate (e.g., `stax` uses a closure), but I can see how storing them together could be convenient.

**Comment by NeilGirdhar** [comment]
I finally decided the shim is too difficult, and went with a slightly more complicated user interface.  Thank you very much for all your help.

**Comment by mattjj** [comment]
I haven't read this thread closely (sorry) but just judging based on the issue title, you might be interested in #4039 which is about to land to allow int primals (associated with a trivial tangent space). We haven't thought through yet how it might help with custom_vjp issues but it might!

**Comment by NeilGirdhar** [comment]
Wow, that's awesome.  That would have solved my problem.
